### PR TITLE
List method for QuestionsController 

### PR DESCRIPTION
**Pull request by HollowPrincess**
https://github.com/malinink/easy-tests/issues/350

- [x] Method
- [x] Test

### CODE REVIEW

**Comment by HollowPrincess** [comment]
@malinink please, have a look: I think cs error is strange and I have no idea, what happend with my controller test

**Comment by HollowPrincess** [comment]
@malinink mission completed

File: src/main/java/easytests/api/v1/controllers/QuestionsController.java [hunk 1]
```diff
@@ -1,33 +1,63 @@
 package easytests.api.v1.controllers;
 
+import easytests.api.v1.exceptions.ForbiddenException;
 import easytests.api.v1.mappers.QuestionsMapper;
-import easytests.core.options.builder.QuestionsOptionsBuilder;
-import easytests.core.services.QuestionsService;
+import easytests.api.v1.models.Question;
+import easytests.common.exceptions.NotFoundException;
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/QuestionsController.java]: use v1

File: src/main/java/easytests/api/v1/controllers/QuestionsController.java [hunk 2]
```diff
@@ -1,33 +1,63 @@
...
+        if (topicModel == null) {
+            throw new NotFoundException();
+        }
+
+        if (!this.acl.hasAccess(topicModel)) {
+            throw new ForbiddenException();
+        }
+
+        final List<QuestionModelInterface> questionsModels =
+                this.questionsService.findByTopic(new TopicModelEmpty(topicId));
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/QuestionsController.java]: use topic model you already have defined

File: src/main/java/easytests/api/v1/mappers/QuestionsMapper.java [hunk 3]
```diff
@@ -9,4 +14,23 @@
...
+    public QuestionsMapper() {
+        super();
+        this.createTypeMap(QuestionModel.class, Question.class)
+            .addMappings(mapper -> {
+                mapper.when(
+                    context -> !(context.getSource() instanceof ModelsListEmpty)
+                ).<Integer>map(questionModel -> questionModel.getQuestionType().getId(),
+                    (question, id) -> question.setType(id));
+                mapper.when(
+                    context -> !(context.getSource() instanceof ModelsListEmpty)
```
Comment (malinink) [inline src/main/java/easytests/api/v1/mappers/QuestionsMapper.java]: there is no need of when condition

File: src/main/java/easytests/api/v1/mappers/QuestionsMapper.java [hunk 4]
```diff
@@ -9,4 +14,23 @@
  */
 @Service("QuestionsMapperV1")
 public class QuestionsMapper extends ModelMapper {
+    public QuestionsMapper() {
+        super();
+        this.createTypeMap(QuestionModel.class, Question.class)
+            .addMappings(mapper -> {
+                mapper.when(
+                    context -> !(context.getSource() instanceof ModelsListEmpty)
```
Comment (malinink) [inline src/main/java/easytests/api/v1/mappers/QuestionsMapper.java]: here also, no need

File: src/main/java/easytests/api/v1/mappers/QuestionsMapper.java [hunk 5]
```diff
@@ -9,4 +14,23 @@
...
+                ).<Integer>map(questionModel -> questionModel.getQuestionType().getId(),
+                    (question, id) -> question.setType(id));
+                mapper.when(
+                    context -> !(context.getSource() instanceof ModelsListEmpty)
+                ).<Integer>map(questionModel -> questionModel.getTopic().getId(),
+                    (question, id) -> question.getTopic().setId(id));
+                mapper.when(
+                    context -> !(context.getSource() instanceof ModelsListEmpty)
```
Comment (malinink) [inline src/main/java/easytests/api/v1/mappers/QuestionsMapper.java]: here too, relation is required

File: src/main/java/easytests/api/v1/models/AdminAnswer.java [hunk 6]
```diff
@@ -10,9 +10,9 @@
 public class AdminAnswer {
     private Integer id;
 
-    private String text;
```
Comment (malinink) [inline src/main/java/easytests/api/v1/models/AdminAnswer.java]: why do you change that?
Add appropriate mapping to fields please.

File: src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java [hunk 7]
```diff
@@ -1,13 +1,193 @@
...
+            final AnswerModel answerModel = new AnswerModel();
+            answerModel.map(answersSupport.getEntityFixtureMock(answerIdx));
+            answersModels.add(answerModel);
+        });
+
+        IntStream.range(0, 2).forEach(idx -> {
+            final QuestionModel questionModel = new QuestionModel();
+            questionModel.map(this.questionSupport.getEntityFixtureMock(idx));
+
+            IntStream.range(0, 2).forEach(answerIdx ->
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java]: no need to set back realtion.

File: src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java [hunk 8]
```diff
@@ -1,13 +1,193 @@
...
+            );
+            questionModel.setAnswers(answersModels);
+            questionsModels.add(questionModel);
+        });
+
+        int topicIdParamValue = 1;
+
+        when(this.topicsService.find(topicIdParamValue))
+                .thenReturn(new TopicModelEmpty(topicIdParamValue));
+        when(this.questionsService.findByTopic(new TopicModelEmpty(topicIdParamValue)))
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java]: better to use any here

File: src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java [hunk 9]
```diff
@@ -1,13 +1,193 @@
...
+                .thenReturn(questionsModels);
+        when(this.answersService.findByQuestion(new QuestionModelEmpty()))
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java]: down thinks there is need of that mocking

File: src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java [hunk 10]
```diff
@@ -1,13 +1,193 @@
...
+                .thenReturn(answersModels);
+        when(this.questionTypesService.find(topicIdParamValue))
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java]: that also, think of what you are trying to test

File: src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java [hunk 11]
```diff
@@ -1,13 +1,193 @@
...
+                                )
+                        ).build()
+                ))
+                .andReturn();
+    }
+
+    @Test
+    public void testListNotFound() throws Exception {
+        int topicIdParamValue = 5;
+
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java:161]: where is mock ?

File: src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java [hunk 12]
```diff
@@ -1,13 +1,193 @@
...
+                .andExpect(status().isNotFound())
+                .andExpect(content().string(""))
+                .andReturn();
+    }
+
+    @Test
+    public void testListForbidden() throws Exception {
+        int topicIdParamValue = 1;
+
+        when(this.topicsService.find(topicIdParamValue))
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java]: strange mock also

**Review by malinink — CHANGES_REQUESTED** [review]

File: src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java (hunk 11 above)
Comment (HollowPrincess) [reply src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java:161]: What mock? I see no list method with mock here

File: src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java (hunk 12 above)
Comment (HollowPrincess) [reply src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java]: Why is it strange?

**Comment by HollowPrincess** [comment]
@malinink fixed

File: src/main/java/easytests/api/v1/controllers/QuestionsController.java [hunk 13]
```diff
@@ -1,33 +1,62 @@
...
+
+        if (topicModel == null) {
+            throw new NotFoundException();
+        }
+
+        if (!this.acl.hasAccess(topicModel)) {
+            throw new ForbiddenException();
+        }
+
+        final List<QuestionModelInterface> questionsModels =
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/QuestionsController.java:59]: you doesn't fetch questions with answers, is that correct?

File: src/main/resources/api/api.yaml [hunk 14]
```diff
@@ -1298,4 +1298,4 @@ definitions:
         type: integer
         format: int32
       message:
-        type: string
+        type: string
```
Comment (malinink) [inline src/main/resources/api/api.yaml]: remove spaces

File: src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java [hunk 15]
```diff
@@ -1,13 +1,183 @@
...
+
+        IntStream.range(0, 2).forEach(idx -> {
+            final QuestionModel questionModel = new QuestionModel();
+            questionModel.map(this.questionSupport.getEntityFixtureMock(idx));
+            questionModel.setAnswers(answersModels);
+            questionsModels.add(questionModel);
+        });
+
+        int topicIdParamValue = 1;
+
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java:105]: It seems to me that you should verify that you select topics with answers here.

File: src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java (hunk 11 above)
Comment (malinink) [reply src/test/java/easytests/api/v1/controllers/QuestionsControllerTest.java:161]: Mock for returning null on fetching question.

**Review by malinink — CHANGES_REQUESTED** [review]

**Comment by HollowPrincess** [comment]
@malinink fixed

File: src/main/java/easytests/api/v1/controllers/QuestionsController.java [hunk 16]
```diff
@@ -1,33 +1,60 @@
...
     @Qualifier("QuestionsMapperV1")
     private QuestionsMapper questionsMapper;
 
-    /**
-     * list
-     */
+    @GetMapping
+    public List<Question> list(@RequestParam(name = "topicId", required = true) Integer topicId)
+            throws NotFoundException, ForbiddenException {
+        final TopicModelInterface topicModel = this.topicsService.find(topicId);
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/QuestionsController.java]: forAuth

**Review by malinink — CHANGES_REQUESTED** [review]

**Review by malinink — APPROVED** [review]

**Pull request merged**